File: src/monid_finance_mcp/receipts.py
from __future__ import annotations

import json
import math
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import TextIO, TypeGuard, cast

from monid_finance_mcp.client import MonidError, MonidRun
from monid_finance_mcp.models import JsonObject, JsonValue

DEFAULT_LEDGER_PATH = Path(__file__).resolve().parents[3] / "receipts" / "ledger.jsonl"


def _utc_now() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def _cost(run: MonidRun | None) -> JsonObject | None:
    if run is None or run.cost is None:
        return None
    return {"value": float(run.cost.value), "currency": run.cost.currency}


def _input_summary(*, body: JsonObject | None, query_params: JsonObject | None) -> JsonObject:
    summary: JsonObject = {}
    if query_params is not None:
        summary["query_params"] = dict(sorted(query_params.items()))
    if body is not None:
        summary["body"] = dict(sorted(body.items()))
    return summary


@dataclass(frozen=True, slots=True)
class Receipt:
    """One measured Monid call attempt, successful or failed."""

    tool: str
    provider: str
    endpoint: str
    run_id: str | None
    lifecycle_status: str | None
    provider_http_status: int | None
    measured_cost: JsonObject | None
    error: str | None
    input_summary: JsonObject

    def to_dict(self) -> JsonObject:
        record: JsonObject = {
            "timestamp": _utc_now(),
            "tool": self.tool,
            "provider": self.provider,
            "endpoint": self.endpoint,
        }
        if self.run_id is not None:
            record["run_id"] = self.run_id
        if self.lifecycle_status is not None:
            record["lifecycle_status"] = self.lifecycle_status
        if self.provider_http_status is not None:
            record["provider_http_status"] = self.provider_http_status
        if self.measured_cost is not None:
            record["measured_cost"] = self.measured_cost
        if self.error is not None:
            record["error"] = self.error
        record["input"] = self.input_summary
        return record


class ReceiptsLedger:
    """Append-only JSONL ledger of every Monid call, kept out of tool responses."""

    def __init__(self, path: Path = DEFAULT_LEDGER_PATH) -> None:
        self._path = path
        self._handle: TextIO | None = None

    def record_success(
        self,
        *,
        tool: str,
        run: MonidRun,
        body: JsonObject | None,
        query_params: JsonObject | None,
    ) -> None:
        self._append(
            Receipt(
                tool=tool,
                provider=run.provider,
                endpoint=run.endpoint,
                run_id=run.run_id,
                lifecycle_status=run.status,
                provider_http_status=run.provider_http_status,
                measured_cost=_cost(run),
                error=None,
                input_summary=_input_summary(body=body, query_params=query_params),
            )
        )

    def record_failure(
        self,
        *,
        tool: str,
        provider: str,
        endpoint: str,
        error: MonidError,
        body: JsonObject | None,
        query_params: JsonObject | None,
    ) -> None:
        run = error.run
        self._append(
            Receipt(
                tool=tool,
                provider=provider,
                endpoint=endpoint,
                run_id=run.run_id if run is not None else None,
                lifecycle_status=run.status if run is not None else None,
                provider_http_status=run.provider_http_status if run is not None else None,
                measured_cost=_cost(run),
                error=f"{type(error).__name__}: {error}",
                input_summary=_input_summary(body=body, query_params=query_params),
            )
        )

    def _append(self, receipt: Receipt) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        if self._handle is None:
            self._handle = self._path.open("a", encoding="utf-8")
        self._handle.write(json.dumps(receipt.to_dict(), sort_keys=False) + "\n")
        self._handle.flush()

    def close(self) -> None:
        if self._handle is not None:
            self._handle.close()
            self._handle = None


def summarize_ledger(path: Path = DEFAULT_LEDGER_PATH) -> JsonObject:
    """Aggregate committed ledger rows for the cost story."""
    tools: dict[str, JsonObject] = {}
    total = 0.0
    calls = 0
    failures = 0
    if path.exists():
        for line in path.read_text(encoding="utf-8").splitlines():
            if not line.strip():
                continue
            loaded: object = json.loads(line)
            if not isinstance(loaded, dict):
                continue
            payload = cast(dict[object, object], loaded)
            row: JsonObject = {}
            for key, value in payload.items():
                if isinstance(key, str) and _is_json_value(value):
                    row[key] = value
            tool = row.get("tool")
            name = tool if isinstance(tool, str) else "unknown"
            calls += 1
            is_failure = row.get("error") is not None
            if is_failure:
                failures += 1
            cost = row.get("measured_cost")
            entry = tools.setdefault(name, {"calls": 0, "failures": 0, "usd_cost": 0.0})
            usd = _usd_value(cost)
            if usd is not None:
                total += usd
                running = _usd_value(entry.get("usd_cost"))
                entry["usd_cost"] = (running or 0.0) + usd
            calls_so_far = _int_value(entry.get("calls"))
            entry["calls"] = calls_so_far + 1
            if is_failure:
                failures_so_far = _int_value(entry.get("failures"))
                entry["failures"] = failures_so_far + 1
    return {
        "calls": calls,
        "failures": failures,
        "total_usd_cost": round(total, 6),
        "tools": dict(sorted(tools.items())),
    }


def _usd_value(cost: JsonValue) -> float | None:
    if not isinstance(cost, dict):
        return None
    currency = cost.get("currency")
    value = cost.get("value")
    if currency != "USD" or isinstance(value, bool) or not isinstance(value, int | float):
        return None
    return float(value)


def _int_value(value: JsonValue) -> int:
    return value if isinstance(value, int) and not isinstance(value, bool) else 0


def _is_json_value(value: object) -> TypeGuard[JsonValue]:
    if value is None or isinstance(value, str | int | bool):
        return True
    if isinstance(value, float):
        return math.isfinite(value)
    if isinstance(value, list):
        items = cast(list[object], value)
        return all(_is_json_value(item) for item in items)
    if isinstance(value, dict):
        mapping = cast(dict[object, object], value)
        return all(isinstance(key, str) and _is_json_value(item) for key, item in mapping.items())
    return False

File: go/fd/receipts.go
// This file ports receipts.py: the append-only JSONL receipts ledger that
// records every Monid call attempt (tool/provider/endpoint/run_id/status/
// measured_cost/error) outside of any tool response, plus its summary
// aggregation.
package fd

import (
	"encoding/json"
	"errors"
	"math"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/belazy/monid-finance/monid"
)

// MeasuredCost is one receipt's measured billing cost, mirroring
// receipts._cost's {"value": ..., "currency": ...} shape.
type MeasuredCost struct {
	Value    float64 `json:"value"`
	Currency string  `json:"currency"`
}

// receiptWire is the on-disk JSONL row shape, in the exact key order
// receipts.Receipt.to_dict emits (timestamp, tool, provider, endpoint,
// then whichever optional fields are present, then input last).
type receiptWire struct {
	Timestamp          string         `json:"timestamp"`
	Tool               string         `json:"tool"`
	Provider           string         `json:"provider"`
	Endpoint           string         `json:"endpoint"`
	RunID              *string        `json:"run_id,omitempty"`
	LifecycleStatus    *string        `json:"lifecycle_status,omitempty"`
	ProviderHTTPStatus *int           `json:"provider_http_status,omitempty"`
	MeasuredCost       *MeasuredCost  `json:"measured_cost,omitempty"`
	Error              *string        `json:"error,omitempty"`
	Input              map[string]any `json:"input"`
}

func utcNowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000000Z")
}

func moneyToMeasuredCost(m *monid.Money) *MeasuredCost {
	if m == nil {
		return nil
	}
	return &MeasuredCost{Value: m.Value, Currency: m.Currency}
}

// inputSummary builds the receipt's "input" field: query_params and/or
// body are included only when the caller passed a non-nil map, mirroring
// receipts._input_summary (which checks `is not None`, not emptiness).
func inputSummary(body, queryParams map[string]any) map[string]any {
	summary := map[string]any{}
	if queryParams != nil {
		summary["query_params"] = queryParams
	}
	if body != nil {
		summary["body"] = body
	}
	return summary
}

// errorLabel renders a call failure the way receipts.Receipt.to_dict
// renders Python's f"{type(error).__name__}: {error}", using the
// monid.RunError's sentinel Kind to pick a label. Unrecognized or plain
// errors fall back to "Error: <message>".
func errorLabel(err error) string {
	if err == nil {
		return "Error: unknown error"
	}
	var runErr *monid.RunError
	if errors.As(err, &runErr) {
		kind := "RunError"
		switch {
		case errors.Is(runErr.Kind, monid.ErrTimeout):
			kind = "TimeoutError"
		case errors.Is(runErr.Kind, monid.ErrProviderHTTP):
			kind = "ProviderHTTPError"
		case errors.Is(runErr.Kind, monid.ErrUnauthorized):
			kind = "UnauthorizedError"
		case errors.Is(runErr.Kind, monid.ErrBlocked):
			kind = "BlockedError"
		case errors.Is(runErr.Kind, monid.ErrSchema):
			kind = "SchemaError"
		case errors.Is(runErr.Kind, monid.ErrNotAllowed):
			kind = "NotAllowedError"
		}
		return kind + ": " + runErr.Message
	}
	return "Error: " + err.Error()
}

// ReceiptsLedger is an append-only JSONL ledger of every Monid call, kept
// out of tool responses, mirroring receipts.ReceiptsLedger.
type ReceiptsLedger struct {
	path     string
	mu       sync.Mutex
	file     *os.File
	spentUSD float64
}

// NewReceiptsLedger returns a ledger that appends to path, creating parent
// directories on first write.
func NewReceiptsLedger(path string) *ReceiptsLedger {
	return &ReceiptsLedger{path: path}
}

// Path returns the ledger's backing file path.
func (l *ReceiptsLedger) Path() string { return l.path }

// SpentUSD returns the cumulative measured USD spend recorded so far.
func (l *ReceiptsLedger) SpentUSD() float64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.spentUSD
}

// RecordSuccess appends one successful-call receipt, mirroring
// ReceiptsLedger.record_success.
func (l *ReceiptsLedger) RecordSuccess(tool string, run *monid.Run, body, queryParams map[string]any) error {
	if run == nil {
		return errors.New("fd: RecordSuccess requires a non-nil run")
	}
	runID := run.RunID
	status := run.Status
	httpStatus := run.ProviderHTTPStatus
	return l.append(receiptWire{
		Timestamp:          utcNowISO(),
		Tool:               tool,
		Provider:           run.Provider,
		Endpoint:           run.Endpoint,
		RunID:              &runID,
		LifecycleStatus:    &status,
		ProviderHTTPStatus: &httpStatus,
		MeasuredCost:       moneyToMeasuredCost(run.Cost),
		Input:              inputSummary(body, queryParams),
	})
}

// RecordFailure appends one failed-call receipt, mirroring
// ReceiptsLedger.record_failure. callErr is typically a *monid.RunError
// (as returned by monid.Client.Run); its Run, if present, supplies run_id/
// lifecycle_status/provider_http_status/measured_cost the same way the
// Python client's exception carries its failed MonidRun.
func (l *ReceiptsLedger) RecordFailure(tool, provider, endpoint string, callErr error, body, queryParams map[string]any) error {
	var runID *string
	var status *string
	var httpStatus *int
	var cost *MeasuredCost
	var runErr *monid.RunError
	if errors.As(callErr, &runErr) && runErr.Run != nil {
		r := runErr.Run
		id, st, hs := r.RunID, r.Status, r.ProviderHTTPStatus
		runID, status, httpStatus = &id, &st, &hs
		cost = moneyToMeasuredCost(r.Cost)
	}
	label := errorLabel(callErr)
	return l.append(receiptWire{
		Timestamp:          utcNowISO(),
		Tool:               tool,
		Provider:           provider,
		Endpoint:           endpoint,
		RunID:              runID,
		LifecycleStatus:    status,
		ProviderHTTPStatus: httpStatus,
		MeasuredCost:       cost,
		Error:              &label,
		Input:              inputSummary(body, queryParams),
	})
}

func (l *ReceiptsLedger) append(rec receiptWire) error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if err := os.MkdirAll(filepath.Dir(l.path), 0o755); err != nil {
		return err
	}
	if l.file == nil {
		f, err := os.OpenFile(l.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
		if err != nil {
			return err
		}
		l.file = f
	}
	raw, err := json.Marshal(rec)
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	if _, err := l.file.Write(raw); err != nil {
		return err
	}
	if rec.MeasuredCost != nil && rec.MeasuredCost.Currency == "USD" {
		l.spentUSD += rec.MeasuredCost.Value
	}
	return nil
}

// Close flushes and closes the ledger's backing file. It is a no-op if
// nothing has been appended yet.
func (l *ReceiptsLedger) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.file == nil {
		return nil
	}
	err := l.file.Close()
	l.file = nil
	return err
}

// ToolSummary is one tool's aggregated ledger stats, mirroring one entry of
// receipts.summarize_ledger's "tools" mapping.
type ToolSummary struct {
	Calls    int     `json:"calls"`
	Failures int     `json:"failures"`
	USDCost  float64 `json:"usd_cost"`
}

// LedgerSummary is the aggregated cost story for a committed ledger,
// mirroring receipts.summarize_ledger's return shape.
type LedgerSummary struct {
	Calls        int                    `json:"calls"`
	Failures     int                    `json:"failures"`
	TotalUSDCost float64                `json:"total_usd_cost"`
	Tools        map[string]ToolSummary `json:"tools"`
}

// SummarizeLedger aggregates the committed ledger rows at path, mirroring
// receipts.summarize_ledger. A missing file summarizes as zero calls.
func SummarizeLedger(path string) (LedgerSummary, error) {
	tools := map[string]ToolSummary{}
	total := 0.0
	calls := 0
	failures := 0

	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return LedgerSummary{Tools: tools}, nil
		}
		return LedgerSummary{}, err
	}
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var row map[string]any
		if err := json.Unmarshal([]byte(line), &row); err != nil {
			continue
		}
		name, _ := row["tool"].(string)
		if name == "" {
			name = "unknown"
		}
		calls++
		_, isFailure := row["error"]
		if isFailure {
			failures++
		}
		entry := tools[name]
		if cost, ok := row["measured_cost"].(map[string]any); ok {
			if currency, _ := cost["currency"].(string); currency == "USD" {
				if value, ok := cost["value"].(float64); ok {
					total += value
					entry.USDCost += value
				}
			}
		}
		entry.Calls++
		if isFailure {
			entry.Failures++
		}
		tools[name] = entry
	}
	return LedgerSummary{
		Calls:        calls,
		Failures:     failures,
		TotalUSDCost: roundTo(total, 6),
		Tools:        tools,
	}, nil
}

func roundTo(v float64, places int) float64 {
	factor := math.Pow(10, float64(places))
	return math.Round(v*factor) / factor
}
